assets/hdfs_append/index.js

```diff
@@ -0,0 +1,161 @@
+    const appendErrors = {};
+
+    const clientService = getClient(context, opConfig, 'hdfs_ha');
+    const hdfsClient = clientService.client;
+
+    function prepareFile(filename, chunks) {
+        // We need to make sure the file exists before we try to append to it.
+        return hdfsClient.getFileStatusAsync(filename)
+            .catch(() => hdfsClient.mkdirsAsync(path.dirname(filename))
+                .then(() => hdfsClient.createAsync(filename, ''))
+                .catch((err) => {
+                    const errMsg = err.stack;
+                    return Promise.reject(
+                        `Error while attempting to create the file: ${filename} on hdfs, error: `
+                        + `${errMsg}`
+                    );
+                }))
+            .return(chunks)
+            // We need to serialize the storage of chunks so we run with concurrency 1
+            .map(chunks, chunk => hdfsClient.appendAsync(filename, chunk), { concurrency: 1 })
+            .catch((err) => {
+                const errMsg = err.stack ? err.stack : err;
+                let sliceError = '';
+                // Detecting the hdfs append error and updating the filename
+                if (errMsg.indexOf('AlreadyBeingCreatedException') > -1) {
+                    let newFilename = '';
+                    if (!appendErrors[filename]) {
+                        newFilename = `${filename}.0`;
+                        appendErrors[filename] = newFilename;
+                    } else {
+                        // Get the last attempted file and increment the number
+                        const incNum = appendErrors[filename].split('.').reverse()[0] * 1 + 1;
+                        newFilename = `${filename}.${incNum}`;
+                        // Set the new target for the next slice attempt
+                        appendErrors[filename] = newFilename;
+                    }
+                    sliceError = `Error sending data to file '${filename}' due to HDFS append `
+                        + `error. Changing destination to '${newFilename}'. Error: ${errMsg}`;
+                } else {
+                    sliceError = `Error sending data to file: ${filename}, error: ${errMsg}`;
+                }
+                if (opConfig.log_data_on_error === true) {
+                    sliceError = `${sliceError} Data: ${JSON.stringify(chunks)}`;
+                }
+                return Promise.reject(sliceError);
+            });
+    }
+
+    return (data) => {
+        // Start by mapping data chunks to their respective files
+        const map = {};
+        data.forEach((record) => {
+            // This skips any records that have non-existant data payloads to avoid empty appends
+            if (record.data.length > 0) {
+                let file = record.filename;
+                // If the file has already had an error, update the filename for the next write
+                // attempt
+                if (appendErrors[file]) {
```

Comment:
Another case where this should be encapsulated in a function to keep the logic here clear.